### PR TITLE
Dev/dplyr

### PR DESCRIPTION
**Pull request by wush978**
I write a smple documentation (97c9bce) and add an environment variable (5d97758ca86d6d8f75eaf5cf0f6d34cfe468dfe1) to enable unit test of dplyr interface on travis-ci.

### CODE REVIEW

**Comment by wush978** [comment]
#48

**Comment by sckott** [comment]
don't worry about the builds failing - the instance we test on I haven't updated recently

File: R/dplyr.R [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+#' my_tbl <- tbl(my_ckan, name = "be2ccdc2-9a76-47bf-862c-60c1525f5b1b")
+#'
+#' # You can use the dplyr verbs with my_tbl. For example:
+#' dplyr::filter(my_tbl, GABARITO == "C")
+#'
+#' }
+#' @aliases dplyr-interface
+#' @export
+src_ckan <- function(url) {
+  library(dplyr)
```
Comment (sckott) [inline R/dplyr.R]: @wush978 can you change this to not use `library(dplyr)`, but instead use something like this https://github.com/ropensci/lawn/blob/65153a64d676e9f7fb5b4b3532bb5d1ac59039a5/R/check_leaflet.R, which checks if the pkg is installed, and if not, stops with message

File: R/dplyr.R [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+#' connect to tables within that CKAN based on the DataStore Data API.
+#'
+#' @param url, the url of the CKAN instance
+#' @examples
+#' \dontrun{
+#' # To connect to a CKAN instance first create a src:
+#' my_ckan <- src_ckan("http://demo.ckan.org")
+#'
+#' # List all tables in the CKAN instance
+#' db_list_tables(src$con)
```
Comment (sckott) [inline R/dplyr.R]: should be `db_list_tables(my_ckan$con)`

File: R/dplyr.R [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    tbl_sql("ckan", src = src, from = sql(from), ...)
+  } else {
+    tbl_sql("ckan", src = src, from = sql(sprintf('SELECT * FROM "%s"', name)))
+  }
+}
+
+#'@export
+src_desc.src_ckan <- function(x) {
+  info <- x$info
+  sprintf("ckan url: %s", src$con@url)
```
Comment (sckott) [inline R/dplyr.R]: should be `sprintf("ckan url: %s", x$con@url)` with `x` instead of `src`

**Comment by sckott** [comment]
I think you need to add `@importFrom` statements for your S3 methods extending from `dplyr`, just like https://github.com/rstats-db/bigrquery/blob/master/R/dplyr.r#L48 - as there's no generic methods to extend unless they are imported from `dplyr`

**Comment by wush978** [comment]
Wow, I don't know that we could import functions and methods from package in `Suggests`. Thanks!

**Comment by sckott** [comment]
well, i guess bigquery has `dplyr` in `Imports` https://github.com/rstats-db/bigrquery/blob/master/DESCRIPTION#L16 - so maybe those `importFrom` statements won't work, though you could try. 

we may have to move `dplyr` to `Imports`

**Comment by wush978** [comment]
Well, after this commit, I get:

```
Namespace dependency not required: ‘dplyr’

See section ‘The DESCRIPTION file’ in the ‘Writing R Extensions’
manual.
* DONE
Status: 1 ERROR
```

during `R CMD check`. So we should either move `dplyr` to `importFrom` or we check the namespace for every related operation.

**Comment by sckott** [comment]
okay, i think we should put in `Imports`

**Comment by wush978** [comment]
OK, I'll modify the DESCRIPTION and eliminate the errors from R CMD check.

**Comment by wush978** [comment]
There is a name conflict:
`fetch` in `DBI.R` and `fetch` in `fetch.R`. They might produce the following warning during `R CMD check`:

```
* checking for code/documentation mismatches ... WARNING
Codoc mismatches from documentation object 'fetch':
fetch
  Code: function(res, n = -1, ...)
  Docs: function(x, store = "session", path = "file", ...)
  Argument names in code not in docs:
    res n
  Argument names in docs not in code:
    x store path
  Mismatches in argument names:
    Position: 1 Code: res Docs: x
    Position: 2 Code: n Docs: store
    Position: 3 Code: ... Docs: path
```

**Comment by sckott** [comment]
We need to have a `fetch` method I assume for the hook into `dplyr` to work? if so, we're still in early versions of the package, so we could change the `fetch()` function name to something else - we could do that after mergin

**Comment by wush978** [comment]
Hmm... it's too late here. I'll check it closely this week.
After no checking error on my machine, I'll push those commits here again.

**Comment by wush978** [comment]
Dear @sckott 

I think `fetch` is required for dplyr:

```
> tb.raw <- dplyr::collect(tb)
Error in (function (classes, fdef, mtable)  : 
  unable to find an inherited method for function ‘fetch’ for signature ‘"CKANResult"’
```

Therefore, we need to rename the `fetch()` function. I can help, but I prefer fixing the unit tests first to prevent any implicit regression of the renaming. How do you think?

**Comment by sckott** [comment]
> I can help, but I prefer fixing the unit tests first to prevent any implicit regression of the renaming. How do you think?

I assume you mean change `ckanr::fetch` not, the one you need for extending dplyr, correct?

Sure, make unit test changes first. 

You're saying we'd change `ckanr::fetch` after the merge?

**Comment by wush978** [comment]
> I assume you mean change ckanr::fetch not, the one you need for extending dplyr, correct?

> Sure, make unit test changes first.

> You're saying we'd change ckanr::fetch after the merge?

Yes, I want to rename the function `ckanr::fetch`. Is there anything I can help about fixing the unit test?

**Comment by sckott** [comment]
> Is there anything I can help about fixing the unit test?

Do you mean the tests running on travis/appveyor?  If so, we have been testing CI against https://data-demo.dpaw.wa.gov.au/ - but not sure why it's not working, possibly some changes on that instance. 

We could change to http://demo.ckan.org/, would the be good?

**Comment by wush978** [comment]
Yes.

I have some progresses about fixing the unit tests tonight. The tests are passed if I use `Test Package` in my Rstudio IDE. However, they failed during `R CMD check` and I cannot figure out where the error is. `testthat` does not display any useful message in the log.

Maybe it will pass in travis. I am checking my work in https://travis-ci.org/wush978/ckanr/builds/120126039

**Comment by wush978** [comment]
After configuring the environment variable in `.travis.yml`,  https://travis-ci.org/wush978/ckanr/builds/120130596 shows:

```
 > test_check("ckanr")

  Error: 404 - Not Found Error

    message Not found

  Execution halted DONE
```

Do you know how to resolve this, @sckott ?

**Comment by sckott** [comment]
@wush978 it looks like your branch `hotfix/test` is branched off of master here, and i assume is behind your `dev/dplyr` branch, did you mean to do that?

**Comment by wush978** [comment]
Yes. I am used to seperate the branches based on their purpose. After submitting `hotfix/test`, I'll merge it into `dev/dplyr`. Does this bother you? I can still directly modify the branch `dev/dplyr`.

**Comment by sckott** [comment]
looks like you've changed the test url in the travis file to https://github.com/ropensci/ckanr/compare/master...wush978:hotfix/test#diff-354f30a63fb0907d4ad57269548329e3R21 but i imagine you'd need to change the other things like dataset id, resource id, etc.

**Comment by wush978** [comment]
I don't understand. On my machine, if I do not set the environment variable, then the `get_test_url()` will return `""` and the test will fail. 

According to the definition of `get_test_url`:

```
> get_test_url
function(){ Sys.getenv("CKANR_TEST_URL") }
<environment: namespace:ckanr>
```

The url is from the environment variable `CKANR_TEST_URL` and there is no such variable in `.travis.yml`. I cannot figure out what the default test url is.

**Comment by sckott** [comment]
The various env vars like `CKANR_TEST_URL` are set as env vars in the Travis web interface, sorry we should have told you that.  You can set all of them in the travis file i think, here's all the env vars 

`CKANR_DEFAULT_KEY`, `CKANR_TEST_KEY`, `CKANR_TEST_GID`, `CKANR_TEST_OID`, `CKANR_TEST_BEHAVIOUR`, `CKANR_TEST_DID`, `CKANR_TEST_RID`, CKANR_TEST_URL`

**Comment by wush978** [comment]
Could I have the value of those non-credential environment variables? Thanks

**Comment by sckott** [comment]
but those are all going to be different if you are using a different CKAN instance - you can set those values to whatever you like

**Comment by wush978** [comment]
Hi sckott,

I just want to update some information. 

I tried to correct some bugs of our unit tests, but the datastore related functions do not work on my machine with <http://demo.ckan.org>.

I cannot search the resource, query the data and insert the dataset with datastore api and I don't know why. For example, an error occurs when I want to create a dataset with datastore api:

```
Browse[2]> content(res, "text")
[1] "{\"help\": \"http://ckan-demo.ckan.io/api/3/action/help_show?name=datastore_create\", \"success\": false, \"error\": {\"records\": [\"The first row is not a json object\"], \"__type\": \"Validation Error\"}}"
```

Do you have any idea?

Furthermore, I will be busy in the following weeks, so I can't resolve this issue soon.

**Comment by sckott** [comment]
Hmm, I don't know why that wouldn't work.  

If okay with you, why don't we merge this. Then we can work on sorting out these few small details, and make the function name change 

We can merge in https://github.com/ropensci/ckanr/compare/master...wush978:hotfix/test?w=1 to master here if you want as well after

**Comment by wush978** [comment]
No problem.

I just open another PR.

**Comment by sckott** [comment]
thanks, merged that one, will fix the ones not passing locally and push back up

**Pull request merged**